encfs/CipherFileIO.cpp

```diff
@@ -326,26 +389,119 @@ int CipherFileIO::generateReverseHeader(unsigned char *headerBuf) {
+        ok = blockRead(tmpReq.data, (int)readSize,
+                       blockNum ^ fileIV);  // cast works because we work on a
+                                            // block and blocksize fit an int
+        if (ok) {
+          // we could have some padding bytes at the end of the plain data (X * 0x00)
+          int padInBytes = 0;
+          while ((padInBytes < (cbs - padOutBytes - 1)) && (tmpReq.data[readSize - padInBytes - 1] == 0x00)) {
+            padInBytes++;
+          }
+          // let's look for the first and needed padding byte, 0x80
+          if (tmpReq.data[readSize - padInBytes - 1] == 0x80) {
+            padInBytes++;
+          } else {
+            padInBytes = 0;
+          }
+          // let's check if fully-sized block is correctly padded
+          if (readSize == bs) {
+            if ((padInBytes >= 1) && (padInBytes <= cbs)) {
+              readSize -= padInBytes;
+            } else {
+              VLOG(1) << "readOneBlock failed (wrong padding) for block " << blockNum << ", size "
+                      << readSize;
+              readSize = -EBADMSG;
+            }
+          // for a partial block, don't return what looks like padding
+          // could be a partially on-disk written block, for whatever reason (write failed, disk full...)
+          } else if ((padOutBytes + padInBytes) <= cbs) {
+            readSize -= padInBytes;
+          }
+        }
+      } else {
+        VLOG(1) << "readSize zero (" << padOutBytes << " padBytes) for offset " << req.offset;
+      }
+    } else if (readSize > 0) {
+      int padBytes = cbs - (readSize % cbs);
+      tmpReq.data[readSize] = 0x80;
+      memset(tmpReq.data + readSize + 1, 0x00, padBytes - 1);
+      readSize += padBytes;
+      ok = blockRead(tmpReq.data, (int)readSize,
+                     blockNum ^ fileIV);  // cast works because we work on a
+                                          // block and blocksize fit an int
+      if (ok) {
+        if ((size_t)readSize < req.dataLen) {
+          // this is the last block, we must pad it with cipherBlockSize() bytes
+          memset(tmpReq.data + readSize, 0x00, cbs - padBytes);
+          readSize += cbs - padBytes;
+        }
+      }
+    } else {
+      // This could be a request for the last padding bytes after the last ciphered block.
+      // we could use getSize() to find out, but as we reverse read, we can assume size
+      // could have changed since our last check, we would then get a wrong cached result.
+      // Let's simply try to read the previous block instead.
+      if (tmpReq.offset > 0) {
+        tmpReq.offset -= tmpReq.dataLen;
+        readSize = base->read(tmpReq);
+      }
+      if ((readSize > 0) && ((bs - readSize) < cbs)) {
+        readSize = cbs - (bs - readSize);
+        memset(tmpReq.data, 0x00, readSize);
+      } else {
+        readSize = 0;
+        VLOG(1) << "readSize zero for offset " << req.offset;
       }
     }
+  }
```

Comment:
Hmm. I feel like this has way too many levels of indentation. Maybe split the subsections into functions? Or at least add comments for what block starts at each "else {" ?